### ISSUE TITLE
Proxy all outgoing application traffic through Bridge to Kubernetes

### ISSUE THREAD

**Issue by jstaf** [microsoft/mindaro#35]
**Is your feature request related to a problem? Please describe.** 

Many environments will not have databases or other dependencies running on Kubernetes or exposed as Kubernetes Services at all. Bridge to Kubernetes does not proxy any traffic that's not linked to a Kubernetes Service. 

Example scenario:

* A user creates a managed database using a service like AWS RDS. The user opts to use a private IP address for the new database (so it's not accessible over the public internet). Once the managed database has been created, the service gives the user an endpoint like `database.laksdjflksadjflk.us-east-1.rds.amazonaws.com` (or even the internal IP address itself like `10.1.2.3`).
* User enters the endpoint of `database.laksdjflksadjflk.us-east-1.rds.amazonaws.com` into an environment variable that is used to configure what database endpoint the app tries to connect to.
* User attempts to launch a development version of their app using "Bridge to Kubernetes". The app crashes spectacularly when the DNS for `database.laksdjflksadjflk.us-east-1.rds.amazonaws.com` either doesn't resolve, or the app is unable to connect to the associated internal ip address.
 
**Describe the solution you'd like**

All outgoing application traffic, DNS lookups, etc. should be proxied to the Kubernetes cluster by Bridge to Kubernetes. This would let the application connect to non-Kubernetes services and endpoints within the Kubernetes cluster network with no additional configuration.
 
**Describe alternatives you've considered**

It's possible to work around this by manually creating a service on the cluster to proxy traffic to your database (or other service external to Kubernetes), and then overriding any env vars or configuration that would reference said database.

```bash
# create externalname service that points to the database
kubectl create service external-name db-proxy-service --external-name database.laksdjflksadjflk.us-east-1.rds.amazonaws.com

# add the following to KubernetesLocalProcessConfig.yaml -
# this will override an example "DBHOST" environment variable that's used by the app to connect to the database
version: 0.1
env:
  - name: DBHOST
    value: $(services:db-proxy-service:5432)
```

This workaround works, but requires two setup steps for every database or non-Kubernetes service, and ideally I'd like to have to not do any setup steps at all.

**Comment by greenie-msft** [comment]
Hi @jstaf 

Thank you for your request and providing the details regarding the scenario you would like to see enabled. I've created a work item to discuss your proposal further with the team. I will respond back to the thread as soon as I have an update to share. 

Thanks.

**Comment by tvdvoorde** [comment]
I can confirm this issue and request. Having Bridge to Kubernetes also proxy outbound traffic from a local process back to the Kubernetes cluster would solve a big issues when developing solutions where backend services have access restrictions.

**Comment by A-S-Blom** [comment]
+1 I can also confirm the issue and feature request. It's also a problem to access services on other namespaces in the k8s cluster.

**Comment by daniv-msft** [comment]
Thanks @tvdvoorde and @A-S-Blom. This issue is tricky, because intercepting all traffic isn't straightforward on all OS and we want a solution that works across Windows, Mac and Linux. We're investigating a potential solution, but this has multiple repercussions that we need to understand/evaluate.

Our approach so far is to ask the developers to declare the dependencies they want to use locally in the [KubernetesLocalProcessConfig.yaml file](https://docs.microsoft.com/visualstudio/containers/configure-bridge-to-kubernetes).
This implies that we support explicitly your scenario there however, which isn't always the case. If your scenario isn't covered, please let us know so that we can evaluate providing support for it.

For example, in your case @A-S-Blom, you should be able to create a `KubernetesLocalProcessConfig.yaml` file at the root path of your project and use this syntax in it to provide access to a service in another namespace when using Bridge:
```
version: 0.1
env:
  - name: MYAPP2_SERVICE_HOST
    value: $(services:myapp2.mynamespace)
```

**Comment by greenie-msft** [comment]
Closing this thread. From what we've seen so far, the KubernetesLocalProcessConfig approach has worked well. Happy to revisit this in the future if we hear more requests for supporting it.

**Comment by anmota-dttl-ie** [comment]
Hi all, I actually have the exactly scenario of the issue creator with the difference only that my DB reside in azure itself. And that his workaround didn't work in my case . And the advice given to use the KubernetesLocalProcessConfig.yaml doesn't even applies to me since the DB is not part of any namespace. 

So it will be great if this issue could be reopened and a real solution implemented. Everything else in this B2K looks excellent.

Cheers.

**Comment by daniv-msft** [comment]
Thanks @anmota-dttl-ie for your comment. Could you please provide more details about your implementation?
Specifically:
* Where is your DB hosted in Azure? As you mention it's not part of a namespace, I assume it's not part of the same AKS cluster, right?
* How is your code calling this DB?

Understanding better your use case would allow us to provide more guidance. The KubernetesLocalProcessConfig.yaml allows Bridge to communicate with other resources hosted in the same Kubernetes cluster, so if you DB isn't part of it this would indeed not be the solution.

**Comment by anmota-dttl-ie** [comment]
Hi, thanks for you quick reply. My setup is pretty standard or so I think... We have a azure Resource Group that contains a K8 cluster with a bunch of K8's in the 'default' namespace that call each others (5 to 8 typically) and then  on the same Resource Group we have other stuff like keyvaults, frontdoors, filestores,  a virtual network and a database server. This is Azure Database for PostgreSQL server.

The K8 services access their db's (a server like this contains just a couple of databases) connecting to it's server name, normally a convention like `name-of-the-resource-group-db.postgres.database.azure.com` plus the usual params db_name, user, password, ssl....

My aim with B2K was to be able to debug and fix code in my local running application to which traffic is redirected but using the remote azure database, not my local database.

I hope this clarifies.

Cheers.

**Comment by tvdvoorde** [comment]
Yes agree with @anmota-dttl-ie 
I was hoping that B2K would fix the common problem when such databases are also restricted to be accessed only by resources (such as AKS) in the same Azure VNET - which prevents developers from accessing them from their local machine.

**Comment by tvdvoorde** [comment]
The work around that @jstaf described will not work for services with TLS that check the hostname

**Comment by daniv-msft** [comment]
Thank you @anmota-dttl-ie and @tvdvoorde for your replies. That really helps!
If if understand properly, we could unblock you by accepting in `KubernetesLocalProcessConfig.yaml` a new `dependencies` value that you could use to set any calls you need Bridge to Kubernetes to proxy from inside the AKS cluster, so that your local machine can call these dependencies:
```
version: 0.1
env:
  - name: MYDEPENDENCY_DATABASE_HOST
    value: $(dependencies:name-of-the-resource-group-db.postgres.database.azure.com)
```

I'm creating a work item on our side to track this work, and reopening this issue so that it's easier to follow up on this.

**Comment by anmota-dttl-ie** [comment]
That looks like it'll do the trick yes... That is really great, thanks a mil for you great support.

Cheers.
